src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp

```diff
@@ -168,22 +168,45 @@ ObjectMetadata HDFSObjectStorage::getObjectMetadata(const std::string & path) co
     return metadata;
 }
 
-void HDFSObjectStorage::listObjects(const std::string & path, RelativePathsWithMetadata & children, size_t max_keys) const
+HDFSFileInfo HDFSObjectStorage::hdfsListDirectoryWrapper(const std::string & path) const
 {
-    initializeHDFSFS();
-    LOG_TEST(log, "Trying to list files for {}", path);
-
     HDFSFileInfo ls;
+
     ls.file_info = hdfsListDirectory(hdfs_fs.get(), path.data(), &ls.length);
+    #if USE_KRB5
+    if (ls.file_info == nullptr && errno == EACCES) // NOLINT
+    {
+        // libhdfs3 wraps GSASL_GSSAPI_INIT_SEC_CONTEXT_ERROR into AccessControlException (EACCES),
+        // which means than sasl context is not active by the reason that krb5 ticket is expired.
+        // runKinit will handle KRB5KRB_AP_ERR_TKT_EXPIRED error and reinitialize ticket again
+        // TODO: it would be good to extract exact gsasl error code from libhdfs3,
+        // but it needs libhdfs3 exceptions refactoring
+        hdfs_builder.runKinit(); // krb5 keytab reinitialization
+        ls.file_info = hdfsListDirectory(hdfs_fs.get(), path.data(), &ls.length);
+    }
+    #endif // USE_KRB5
 
     if (ls.file_info == nullptr && errno != ENOENT) // NOLINT
     {
         // ignore file not found exception, keep throw other exception,
         // libhdfs3 doesn't have function to get exception type, so use errno.
-        throw Exception(ErrorCodes::ACCESS_DENIED, "Cannot list directory {}: {}",
+        if (ls.file_info == nullptr && errno != ENOENT) // NOLINT
+        {
+            throw Exception(ErrorCodes::ACCESS_DENIED, "Cannot list directory {}: {}",
                         path, String(hdfsGetLastError()));
+        }
     }
 
+    return ls;
+}
+
+void HDFSObjectStorage::listObjects(const std::string & path, RelativePathsWithMetadata & children, size_t max_keys) const
+{
+    initializeHDFSFS();
+    LOG_TEST(log, "Trying to list files for {}", path);
+
+    HDFSFileInfo ls = hdfsListDirectoryWrapper(path);
```

Comment:
Why do we care about listObjects only?
Is not it possible to have this issues during another call?

Reply:
I found 3 types of operations with hdfs engine:
- create
  Checked `create` - it resets krb5 by itself
- select
  It was initial issue - I was the main point I can check
- insert
  I just did not meet any complains about it and do not have env to check it

That is why I decided to make a  minimal fix for the moment. If I understand the problem correct - it can affect not hdfs only, but any sasl interaction, for example kafka and maybe other things, but the change could be too global

Reply:
Can ReadBufferFromHDFS be affected?

Just asking, not imply any action.

Reply:
IDK. Just an idea that during long copy to another table it could happen, but during my tests - minimal possible expiration time was 3m and I did not catch it